cmds/tail/tail.go

```diff
@@ -0,0 +1,243 @@
+		}
+		var (
+			foundLines = 0
+			idx        int
+		)
+		for {
+			if foundLines >= numLines {
+				break
+			}
+			// find newlines backwards from the end of `slice`
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	_, err = writer.Write(data)
+	if err != nil {
+		return err
+	}
+	// reposition the stream at the end of what we've read
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
+	return nil
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	if numLines < 0 {
+		return fmt.Errorf("The number of lines cannot be negative")
+	}
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
+	if err != nil {
+		return err
+	}
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
+	var slice []byte             // will hold the final data, after moving line by line
+	for {
+		n, err := io.ReadFull(&inFile, buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
+				return err
+			}
+		}
+		// look for newlines and keep a slice starting at the n-th to last line
+		// (no further than numLines)
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
+		foundNewLines := bytes.Count(slice, []byte{'\n'})
+		newLinesToSkip := 0
+		if foundNewLines > numLines {
+			newLinesToSkip = foundNewLines - numLines
+		}
+		// if there are more than `numLines` lines, skip them
+		for idx := 0; idx < newLinesToSkip; idx++ {
+			pos := bytes.IndexByte(slice, '\n')
+			// in theory the code below should never panic for out of bounds because
+			// of the checks above, but..
+			if len(slice) < pos+1 {
+				return fmt.Errorf("less bytes than expected were found")
+			}
+			slice = slice[pos+1:]
+		}
+	}
+	_, err = writer.Write(slice)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func Tail(inFile os.File, writer io.Writer, follow bool, numLines int) error {
+	var err error
+	if inFile == *os.Stdin {
```

Comment:
So in general this test is not sufficient. What if they give the name /dev/fd/0? That's stdin. What about all the other character devices? 

 If you really want to do this you need to stat the file, see if it's a block device, then call your backwards. 

Or, simpler, see if seeking from the end of tne file works and, if so, read backwards. Maybe a thing is seekable even if it's not a block device.

Reply:
very valid point. I will change this to try to read backwards and fall back to forward reading if that fails